In [4]:
import requests
from datetime import datetime
import pandas as pd
import time

In [17]:
page_number = 1
year = 2024
primary_release_date_gte = f"{year}-01-01"
primary_release_date_lte = datetime.now().strftime("%Y-%m-%d")


url_all_movies = (
    f"https://api.themoviedb.org/3/discover/movie?include_adult=false&include_video=true&language=fr-FR&page={page_number}&sort_by=primary_release_date.desc&primary_release_date.gte={primary_release_date_gte}&primary_release_date.lte={primary_release_date_lte}"
)

headers = {
    "accept": "application/json",
    "Authorization": "Bearer eyJhbGciOiJIUzI1NiJ9.eyJhdWQiOiIxOWYxZmM4MjllODVmZjRiMzg4ZDJiZDYyZTkyZDU2OCIsIm5iZiI6MTc0NzgxMjIxNC4yMTIsInN1YiI6IjY4MmQ3Zjc2OWQwNzg5ZGZiMDhjMTA1NSIsInNjb3BlcyI6WyJhcGlfcmVhZCJdLCJ2ZXJzaW9uIjoxfQ.d-Qt3ABIvB3iqZxWBhR5vvkc5cKkOuH9HeYsc2yFaTk"
}


response = requests.get(url_all_movies, headers=headers)

In [ ]:
display(response.json())

In [ ]:
# On va devoir récuperer les colonnes qui nous manquent pour l'affichage:
# affiche / bande-annonce / classification 


In [7]:
# Récuperer les videos:
# https://api.themoviedb.org/3/movie/{movie_id}/videos?language=fr-FR"

link_ytb = "https://www.youtube.com/watch?v="

def get_movie_videos(movie_id: int):
    """
    Récupère les vidéos associées à un film donné par son ID.
    """
    # URL pour récupérer les vidéos du film
    url = f"https://api.themoviedb.org/3/movie/{movie_id}/videos?language=fr-FR"
    response = requests.get(url, headers=headers)

    # Si on a une réponse, on garde uniquement la premiere vidéo qui correspond à la bande annonce la plus récente
    if response.status_code == 200:
        data = response.json()
        results = data.get('results', [])
        for ba in results:
            if ba.get('official') and ba.get('site') == 'YouTube' and ba.get('type') == 'Trailer':
                return link_ytb + ba.get('key')
        print(f"Aucune bande annonce trouvée pour le film ID {movie_id}.")
    else:
        print(f"Erreur lors de la récupération des vidéos pour le film ID {movie_id}: {response.status_code}")
        return []
    
    
    

In [16]:
# Récuperer des id:
# https://api.themoviedb.org/3/movie/{movie_id}/external_ids

def get_movie_ids(movie_id: int):
    """
    Récupère les IDs externes d'un film spécifique.
    """
    url = f"https://api.themoviedb.org/3/movie/{movie_id}/external_ids"
    response = requests.get(url, headers=headers)

    if response.status_code == 200:
        data = response.json()
        return data.get('imdb_id')
    else:
        print(f"Erreur lors de la récupération des IDs externes pour le film {movie_id}: {response.status_code}")
        return None

In [20]:
df_movies = get_movies(page_count=100, primary_release_date_gte=primary_release_date_gte, primary_release_date_lte=primary_release_date_lte)
print(primary_release_date_gte, primary_release_date_lte)
df_movies

2024-01-01 2025-06-04


,id,poster_path,video,tconst,frenchOverview
0,1137350,https://image.tmdb.org/t/p/original//aWucFFSKS...,https://www.youtube.com/watch?v=YuZfbZR2So8,tt30840798,Zsa-Zsa Korda est un richissime homme d’affair...
1,1001414,https://image.tmdb.org/t/p/original//8HWIML6j4...,https://www.youtube.com/watch?v=S645vaJ0CEk,tt31433402,"Shadyside, 1988. Qui sera reine de promo au ba..."
2,1098006,https://image.tmdb.org/t/p/original//yBpFCMnoS...,https://www.youtube.com/watch?v=n6wZkA4Cb1w,tt27075958,Un chasseur de trésors chevronné rassemble une...
3,575265,https://image.tmdb.org/t/p/original//AozMgdALZ...,https://www.youtube.com/watch?v=Si8mScZ-RgY,tt9603208,Ethan Hunt se rend à Londres avec son équipe d...
4,552524,https://image.tmdb.org/t/p/original//71IjwRa88...,https://www.youtube.com/watch?v=fBYp1DtdHuE,tt11655566,L’histoire touchante et drôle d’une petite fil...


{
  "page": 1,
  "results": [
    {
      "adult": false,
      "backdrop_path": "/7Zx3wDG5bBtcfk8lcnCWDOLM4Y4.jpg",
      "genre_ids": [
        10751,
        35,
        878
      ],
      "id": 552524,
      "original_language": "en",
      "original_title": "Lilo & Stitch",
      "overview": "L’histoire touchante et drôle d’une petite fille hawaïenne solitaire et d’un extra-terrestre fugitif qui l’aide à renouer le lien avec sa famille.",
      "popularity": 679.0045,
      "poster_path": "/71IjwRa88OJMYJBntId7nn0eFHy.jpg",
      "release_date": "2025-05-17",
      "title": "Lilo & Stitch",
      "video": false,
      "vote_average": 7.1,
      "vote_count": 427
    },

In [25]:
def complete_movie_data(imdb_id: str):
    """Récupere l'affiche, la bande annonce et la classification d'un film par son ID.

    Args:
        id (str): tconst du film à récupérer.
    Returns:
        dict: Dictionnaire contenant l'affiche, la bande annonce et la classification du film.
    """
    url = f"https://api.themoviedb.org/3/find/{imdb_id}?external_source=imdb_id&language=fr-FR"
    response = requests.get(url, headers=headers)
    if response.status_code == 200:
        data = response.json()
        if data['movie_results'][0]:
            return {
                'id': data['movie_results'][0].get('id'),
                'poster_path': poster_url + data['movie_results'][0].get('poster_path', ''),
                'video': get_movie_videos(data['movie_results'][0].get('id')),
                'frenchOverview': data['movie_results'][0].get('overview', ''),
            }
    else:
        print(f"Erreur lors de la récupération des données pour l'ID {imdb_id}: {response.status_code}")
        return {
            'id': None,
            'poster_path': None,
            'video': None,
            'frenchOverview': None
        }
        

In [27]:
complete_movie_data('tt0006864')

Aucune bande annonce trouvée pour le film ID 3059.


{'id': 3059,
 'poster_path': 'https://image.tmdb.org/t/p/original//t272DJA8hoq7Ru0BikCWi6GYi26.jpg',
 'video': None,
 'frenchOverview': "À travers l'image d'une femme berçant un enfant, quatre épisodes de l'intolérance sont racontés dans une fresque monumentale. Un épisode moderne sur un gréviste condamné à la pendaison. Un épisode biblique lors d'une noce à Cana. Un épisode des guerres de religion au temps de Charles IX. Un épisode chaldéen. Dans tous l'intolérance l'emporte sur l'amour."}

In [34]:
# Récuperer tous les films manquants dans notre dataframe pre machine learning
df = pd.read_csv('../../ressources/df_final.csv', sep=';', index_col=0)
df_copy = df.copy()
df_copy = df.reset_index()

#df_v4['decade'] = (df_v4['startYear'] // 10) * 10

df_copy = df_copy.drop(columns=['primaryTitle', 'title', 'primaryTitle', 'production_countries', 'runtimeMinutes', 'revenue', 'tagline', 'overview', 'id'], axis=1)
df_copy 

,tconst,frenchTitle,startYear,genres,averageRating,numVotes,actor1,actor2,actor3,directors
0,tt0006864,Intolérance,1916,"History, Drama",7.7,17434,Lillian Gish,Robert Harron,Mae Marsh,D.W. Griffith
1,tt0009968,Le lys brisé,1919,"Drama, Romance",7.2,11483,Lillian Gish,Richard Barthelmess,Donald Crisp,D.W. Griffith
2,tt0010307,J'accuse,1919,"History, War, Drama, Horror, Romance",7.7,2249,Romuald Joubé,Maxime Desjardins,Séverin-Mars,Abel Gance
3,tt0010418,L'admirable Crichton,1919,"Drama, Adventure, Fantasy",7.0,2032,Thomas Meighan,Theodore Roberts,Raymond Hatton,Cecil B. DeMille
4,tt0011439,Le signe de Zorro,1920,"Adventure, Drama, Western, Action, Romance",7.0,2946,Douglas Fairbanks,Marguerite De La Motte,Noah Beery,Fred Niblo
...,...,...,...,...,...,...,...,...,...,...
17266,tt9902160,Herself,2020,Drama,7.0,5127,Molly McCann,Clare Dunne,Ruby Rose O'Hara,Phyllida Lloyd
17267,tt9904802,Enemy Lines,2020,"Drama, Action, War",4.6,2045,Ed Westwick,John Hannah,Tom Wisdom,Anders Banke
17268,tt9907782,Eight for Silver,2021,"Mystery, Fantasy, Horror",6.2,20425,Boyd Holbrook,Kelly Reilly,Alistair Petrie,Sean Ellis
17269,tt9908390,Le lion,2020,Comedy,5.5,1522,Dany Boon,Philippe Katerine,Anne Serra,Ludovic Colbeau-Justin


In [6]:
# Fonction qui va récuperer les films en fonction de la date de sortie et du nombre de votes
def get_movies(page_count: int, primary_release_date_gte: str, primary_release_date_lte: str):
    all_movies_data = []
    for page_number in range(1, page_count + 1):
        url = (
            f"https://api.themoviedb.org/3/discover/movie"
            f"?include_adult=false&include_video=true&language=fr-FR"
            f"&page={page_number}&sort_by=vote_count.desc"
            f"&primary_release_date.gte={primary_release_date_gte}"
            f"&primary_release_date.lte={primary_release_date_lte}"
        )
        response = requests.get(url, headers=headers)
        if response.status_code == 200:
            data = response.json()
            movies = data.get('results', [])
            for movie in movies:
                if movie.get('original_language') in ['fr', 'en'] and movie.get('vote_count', 0) >= 500:
                    all_movies_data.append(movie.get('id'))
        else:
            print(f"Erreur lors de la récupération des données pour la page {page_number}: {response.status_code}")
            break
    return all_movies_data

In [9]:
# Récuperer les details des films pour notre df de ML et Display
def get_movie_details(movie_id):
    url = f"https://api.themoviedb.org/3/movie/{movie_id}?language=fr-FR&append_to_response=videos,credits"
    response = requests.get(url, headers=headers)
    time.sleep(0.25)  # Respect du rate limit
    if response.status_code == 200:
        data = response.json()
        # Récupération des réalisateurs
        directors = [crew['name'] for crew in data.get('credits', {}).get('crew', []) if crew.get('job') == 'Director']
        # Récupération des 3 premiers acteurs
        actors = [cast['name'] for cast in data.get('credits', {}).get('cast', [])][:3]
        # Récupération de la bande-annonce (YouTube)
        video = None
        for v in data.get('videos', {}).get('results', []):
            if v.get('type') == 'Trailer' and v.get('site') == 'YouTube':
                video = f"https://www.youtube.com/watch?v={v.get('key')}"
                break
        # Récupération des genres
        genres = [g['name'] for g in data.get('genres', [])]
        return {
            'frenchTitle': data.get('title'),
            'primaryTitle': data.get('original_title'),
            'averageRating': data.get('vote_average'),
            'runtimeMinutes': data.get('runtime'),
            'genres': ', '.join(genres),
            'startYear': int(data.get('release_date', '0000')[:4]) if data.get('release_date') else None,
            'overview': data.get('overview'),
            'directors': ', '.join(directors),
            'acteurs': ', '.join(actors),
            'poster_path': f"https://image.tmdb.org/t/p/w500{data.get('poster_path')}" if data.get('poster_path') else None,
            'video': video,
            'frenchOverview': data.get('overview'),  # ou autre champ si tu veux une version différente
        }
    else:
        print(f"Erreur détails film ID {movie_id} : code {response.status_code}")
        return None

In [18]:
results = get_movies(page_count=500, primary_release_date_gte=primary_release_date_gte, primary_release_date_lte=primary_release_date_lte)

details_movie = []

for id in results:
    movie_details = get_movie_details(id)
    if movie_details:
        details_movie.append(movie_details)
df_movies = pd.DataFrame(details_movie)
display(df_movies)

,frenchTitle,primaryTitle,averageRating,runtimeMinutes,genres,startYear,overview,directors,acteurs,poster_path,video,frenchOverview
0,Deadpool & Wolverine,Deadpool & Wolverine,7.595,128,"Action, Comédie, Science-Fiction",2024,Un Wade Wilson désabusé s'abrutit de travail d...,Shawn Levy,"Ryan Reynolds, Hugh Jackman, Emma Corrin",https://image.tmdb.org/t/p/w500/znq3nu6wbSha6c...,https://www.youtube.com/watch?v=KT-8QL04x40,Un Wade Wilson désabusé s'abrutit de travail d...
1,Dune : Deuxième partie,Dune: Part Two,8.146,165,"Science-Fiction, Aventure",2024,Le voyage mythique de Paul Atreides qui s'alli...,Denis Villeneuve,"Timothée Chalamet, Zendaya, Rebecca Ferguson",https://image.tmdb.org/t/p/w500/rBrJl1vUtaM6NU...,https://www.youtube.com/watch?v=oGimpDBwkFc,Le voyage mythique de Paul Atreides qui s'alli...
2,Vice-versa 2,Inside Out 2,7.554,97,"Animation, Aventure, Comédie, Familial",2024,"Riley, adolescente depuis peu, au moment où le...",Kelsey Mann,"Amy Poehler, Maya Hawke, Kensington Tallman",https://image.tmdb.org/t/p/w500/9Jic8z9NEeCEmS...,https://www.youtube.com/watch?v=9xoCn6D-PIw,"Riley, adolescente depuis peu, au moment où le..."
3,Le Robot sauvage,The Wild Robot,8.300,102,"Animation, Science-Fiction, Familial",2024,L’incroyable épopée d'un robot -- l'unité ROZZ...,Chris Sanders,"Lupita Nyong'o, Pedro Pascal, Kit Connor",https://image.tmdb.org/t/p/w500/2IUCa73cvvIgZS...,https://www.youtube.com/watch?v=3iHLk8nHX80,L’incroyable épopée d'un robot -- l'unité ROZZ...
4,The Substance,The Substance,7.143,141,"Horreur, Science-Fiction",2024,"Elisabeth Sparkle, vedette d’une émission d’aé...",Coralie Fargeat,"Demi Moore, Margaret Qualley, Dennis Quaid",https://image.tmdb.org/t/p/w500/5Jfj6LqxOludXg...,https://www.youtube.com/watch?v=LbOqfmP-Pb8,"Elisabeth Sparkle, vedette d’une émission d’aé..."
...,...,...,...,...,...,...,...,...,...,...,...,...
139,Salem's Lot,Salem's Lot,6.100,113,Horreur,2024,Ben Mears revient dans sa ville natale de Sale...,Gary Dauberman,"Lewis Pullman, Makenzie Leigh, Jordan Preston ...",https://image.tmdb.org/t/p/w500/YFD4QcyNREbq2e...,None,Ben Mears revient dans sa ville natale de Sale...
140,Scandaleusement vôtre,Wicked Little Letters,6.939,102,"Comédie, Drame, Mystère",2024,"Lorsque les habitants de Littlehampton, y comp...",Thea Sharrock,"Olivia Colman, Jessie Buckley, Anjana Vasan",https://image.tmdb.org/t/p/w500/3OB8B8iSznL6rl...,https://www.youtube.com/watch?v=eXoogFVnFko,"Lorsque les habitants de Littlehampton, y comp..."
141,G20,G20,6.600,108,"Action, Mystère, Drame",2025,"Lorsque le sommet du G20 est pris d'assaut, la...",Patricia Riggen,"Viola Davis, Anthony Anderson, Marsai Martin",https://image.tmdb.org/t/p/w500/tDAG6YqKHJYZJt...,https://www.youtube.com/watch?v=z7BgGFK3EBE,"Lorsque le sommet du G20 est pris d'assaut, la..."
142,Mothers' Instinct,Mothers' Instinct,6.900,94,"Thriller, Drame",2024,"Au début des années 60, deux voisines, Alice e...",Benoît Delhomme,"Anne Hathaway, Jessica Chastain, Anders Daniel...",https://image.tmdb.org/t/p/w500/wIDulcEDrbW5tL...,None,"Au début des années 60, deux voisines, Alice e..."
